buildSrc/src/main/kotlin/Java9Modularity.kt

```diff
@@ -0,0 +1,65 @@
+import org.gradle.api.JavaVersion
+import org.gradle.api.Project
+import org.gradle.api.tasks.bundling.Jar
+import org.gradle.kotlin.dsl.findByType
+import org.jetbrains.kotlin.gradle.dsl.KotlinMultiplatformExtension
+import org.jetbrains.kotlin.gradle.targets.jvm.KotlinJvmTarget
+import java.util.spi.ToolProvider
+
+object Java9Modularity {
+    @JvmStatic
+    fun Project.configureJava9ModuleInfo(moduleName: String) {
+        val multiplatform = extensions.findByType<KotlinMultiplatformExtension>() ?: return
+        val jvmTargets = multiplatform.targets.filterIsInstance<KotlinJvmTarget>().filter { it.withJavaEnabled }
+        if (jvmTargets.isEmpty()) {
+            logger.warn("No Kotlin JVM targets found with Java enabled, can't configure Java 9 source set!")
+        }
+        jvmTargets.forEach { target ->
+            val mainCompilation = target.compilations.getByName("main")
+            val inputClasspath = mainCompilation.compileKotlinTask.classpath
+            val outputClassesDir = mainCompilation.compileKotlinTask.destinationDir
+            val outputJarTask = tasks.getByName(target.artifactsTaskName) as Jar
+
+            val java9SourceSet = multiplatform.sourceSets.create("${target.targetName}Java9")
+            val java9Compilation = target.compilations.create("java9") {
+                compileJavaTaskProvider!!.configure {
+                    dependsOn(mainCompilation.compileAllTaskName)
+                    source(java9SourceSet.kotlin)
+                    sourceCompatibility = JavaVersion.VERSION_1_9.toString()
+                    targetCompatibility = JavaVersion.VERSION_1_9.toString()
+                    doFirst {
+                        classpath = files()
+                        options.compilerArgs = listOf(
+                            "--release", "9",
+                            "--module-path", inputClasspath.asPath,
+                            "--patch-module", "$moduleName=$outputClassesDir",
+                            "-Xlint:-requires-transitive-automatic"
+                        )
+                    }
+                }
+            }
+
+            outputJarTask.apply {
+                from(java9Compilation.output.classesDirs) {
+                    into("META-INF/versions/9")
+                }
+                manifest {
+                    attributes(mapOf("Multi-Release" to true))
+                }
+            }
+
+            tasks.register("${target.targetName}CheckModuleInfo") {
+                dependsOn(outputJarTask)
+                doLast {
+                    val jdeps = ToolProvider.findFirst("jdeps").orElseThrow { IllegalStateException("Tool 'jdeps' is not available") }
+                    jdeps.run(
```

Comment:
It would be nice if we could detect if `jdeps` reports different dependencies from the current module-info and fail the task. This is needed to check automatically on CI that no new dependencies are accidentally introduced.

However, this may block us from removing some dependencies manually

Reply:
I also don't see task to 'write' module-info. But from your comment I assume the same `CheckModuleInfo` task can be invoked and results copied from console, right?

Reply:
To automatically check differences I could capture the (console) output that comes from `jdeps` and throw an exception if something along the lines of `[Suggested module descriptor for xxx]` is in the output. If you remove dependencies you would need to remove the appropriate `requires` line from the module-info too, which I believe to be the desired effect.

For 'writing' the module-info you would indeed need to copy the suggested module-descriptor manually from the console output into the repository version of the module-info.